apps/prairielearn/src/lib/authz-data.ts

```diff
@@ -1,156 +1,199 @@
+  user,
   course_id,
   course_instance_id,
-  is_administrator,
   ip,
   req_date,
-  req_mode = null,
+  overrides,
 }: {
-  authn_user: User;
+  user: User;
   course_id: string | null;
   course_instance_id: string | null;
-  is_administrator: boolean;
   ip: string | null;
   req_date: Date;
-  req_mode?: string | null;
-}) {
+  overrides: {
+    is_administrator: boolean;
+    req_mode?: EnumMode;
+    req_course_role?: EnumCourseRole;
+    req_course_instance_role?: EnumCourseInstanceRole;
+    allow_example_course_override?: boolean;
+  };
+}): Promise<ConstructedCourseOrInstanceContext> {
+  const resolvedOverrides = {
+    allow_example_course_override: true,
+    ...overrides,
+  };
+  assert(course_id !== null || course_instance_id !== null);
+
   const isCourseInstance = Boolean(course_instance_id);
 
-  const rawAuthzData = await selectAuthzData({
-    user_id: authn_user.user_id,
+  const rawAuthzData = await selectCourseOrInstanceContextData({
+    user_id: user.user_id,
     course_id,
     course_instance_id,
-    is_administrator,
-    allow_example_course_override: true,
     ip,
     req_date,
-    req_mode,
-    req_course_role: null,
-    req_course_instance_role: null,
   });
 
   if (rawAuthzData === null) {
     return {
       authzData: null,
-      authzCourse: null,
-      authzInstitution: null,
-      authzPermissionsCourse: null,
-      authzCourseInstance: null,
+      course: null,
+      institution: null,
+      courseInstance: null,
     };
   }
 
-  const permissions_course = rawAuthzData.permissions_course;
+  const course_role = run(() => {
+    if (resolvedOverrides.req_course_role != null) {
+      return resolvedOverrides.req_course_role;
+    }
+    if (resolvedOverrides.is_administrator) {
+      return 'Owner';
+    }
+
+    if (rawAuthzData.course.example_course) {
+      // If the course is an example course and the override is allowed, return Viewer.
+      if (
+        resolvedOverrides.allow_example_course_override &&
+        // If we can step _up_ to Viewer, do so.
+        // We don't want to accidentally decrease the role of an existing user.
+        ['None', 'Previewer'].includes(rawAuthzData.permissions_course.course_role)
+      ) {
+        return 'Viewer';
+      }
+
+      // Otherwise, return the actual role.
+      return rawAuthzData.permissions_course.course_role;
+    }
+
+    return rawAuthzData.permissions_course.course_role;
+  });
+
+  const course_instance_role = run(() => {
+    if (resolvedOverrides.req_course_instance_role != null) {
+      return resolvedOverrides.req_course_instance_role;
+    }
+    if (resolvedOverrides.is_administrator) {
+      return 'Student Data Editor';
+    }
+    return rawAuthzData.permissions_course_instance.course_instance_role;
+  });
+
+  const mode = run(() => {
+    if (resolvedOverrides.req_mode != null) {
+      return resolvedOverrides.req_mode;
+    }
+    return rawAuthzData.mode;
+  });
```

Comment:
Nit, not that it really matters, but I think nullish coalescing is a little cleaner here:

```suggestion
  const mode = resolvedOverrides.req_mode ?? rawAuthzData.mode
```